File: plugins/collector/event/apps/container/docker.go
package container

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Docker struct {
	version string
	reg     *regexp.Regexp
}

func (Docker) Name() string { return "docker" }

func (Docker) Type() string { return "container" }

func (d Docker) Version() string { return d.version }

func (Docker) Match(p *process.Process) bool { return p.Name == "dockerd" }

func (n *Docker) Run(p *process.Process) (m map[string]string, err error) {
	if n.reg == nil {
		n.reg = regexp.MustCompile(`Docker\sversion\s(\d+\.)+\d+`)
	}
	result, err := apps.Execute(p, "--version")
	str := n.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	n.version = strings.TrimPrefix(str, "Docker version ")
	return
}

func init() { apps.Regist(&Docker{}) }

File: plugins/collector/src/event/apps/container/docker.rs
use crate::event::apps::{execute, AppProc, IApp};
use std::collections::HashMap;

pub struct Docker {
    version: String,
}

impl Docker {
    pub fn new() -> Self { Docker { version: String::new() } }
}

impl IApp for Docker {
    fn name(&self)     -> &'static str { "docker" }
    fn app_type(&self) -> &'static str { "container" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "dockerd" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "Docker version 20.10.18, build …"
        if let Some(rest) = out.strip_prefix("Docker version ") {
            self.version = rest.split([',', ' ']).next().unwrap_or("").to_owned();
        }
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
